theano/gof/cc.py

```diff
@@ -1512,16 +1622,40 @@ def get_dynamic_module(self):
         """ % dict(struct_name=self.struct_name)
 
         # We add all the support code, compile args, headers and libs we need.
-            for support_code in self.support_code() + self.c_support_code_apply:
-                mod.add_support_code(support_code)
-            mod.add_support_code(self.struct_code)
-            mod.add_support_code(static)
-            mod.add_function(instantiate)
-            for header in self.headers():
-                mod.add_include(header)
-            for init_code_block in self.init_code() + self.c_init_code_apply:
-                mod.add_init_code(init_code_block)
-            self._mod = mod
+                for support_code in self.support_code() + self.c_support_code_apply:
+                    mod.add_support_code(support_code)
+
+                    if not self.c_callable:
+                        mod.add_support_code("""
+                                             #ifdef _WIN32
+                                             #define DllExport __declspec(dllexport)
+                                             #else
+                                             #define DllExport
+                                             #endif
+                                             """)
+                        mod.add_support_code(self.struct_code)
+                    else:
+                        mod.add_header_code("""
+                                            #ifdef _WIN32
+                                            #define DllExport __declspec(dllexport)
+                                            #else
+                                            #define DllExport
+                                            #endif
+                                            """)
+                        mod.add_header_code(self.struct_code)
```

Comment:
For a c_callable function the declaration should be added to the header code of the shared library while for a 'normal' function it should be added directly to the *.cpp module.